pkg/common/telemetry/metrics_benchmark_test.go

```diff
@@ -0,0 +1,148 @@
+package telemetry
+
+import (
+	"testing"
+	"time"
+)
+
+var (
+	key    = []string{"key1", "key2", "key3"}
+	valf   = float32(5.0)
+	labels = []Label{
+		{
+			Name:  "lkey1",
+			Value: "lval1",
+		},
+		{
+			Name:  "lkey2",
+			Value: "lval2",
+		},
+	}
+	valt = time.Now()
+)
+
+func BenchmarkDogStatsd(b *testing.B) {
+	m := getDogStatsdMetricImpl(b)
+
+	benchmarkMetricImpl(b, m)
+}
+
+func BenchmarkInMem(b *testing.B) {
+	m := getInMemMetricImpl(b)
+
+	benchmarkMetricImpl(b, m)
+}
+
+func BenchmarkM3(b *testing.B) {
+	m := getM3MetricImpl(b)
+
+	benchmarkMetricImpl(b, m)
+}
+
+func BenchmarkPrometheus(b *testing.B) {
+	m := getPrometheusMetricImpl(b)
+
+	benchmarkMetricImpl(b, m)
+}
+
+func BenchmarkStatsd(b *testing.B) {
+	m := getStatsdMetricImpl(b)
+
+	benchmarkMetricImpl(b, m)
+}
+
+func benchmarkMetricImpl(b *testing.B, m Metrics) {
+	b.Run("SetGauge", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.SetGauge(key, valf)
+		}
+	})
+
+	b.Run("SetGaugeWithLabels", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.SetGaugeWithLabels(key, valf, labels)
+		}
+	})
+
+	b.Run("EmitKey", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.EmitKey(key, valf)
+		}
+	})
+
+	b.Run("IncrCounter", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.IncrCounter(key, valf)
+		}
+	})
+
+	b.Run("IncrCounterWithLabels", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.IncrCounterWithLabels(key, valf, labels)
+		}
+	})
+
+	b.Run("AddSample", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.AddSample(key, valf)
+		}
+	})
+
+	b.Run("AddSampleWithLabels", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.AddSampleWithLabels(key, valf, labels)
+		}
+	})
+
+	b.Run("MeasureSince", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.MeasureSince(key, valt)
+		}
+	})
+
+	b.Run("MeasureSinceWithLabels", func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			m.MeasureSinceWithLabels(key, valt, labels)
+		}
+	})
+}
+
+func getDogStatsdMetricImpl(b *testing.B) Metrics {
+	m, err := NewMetrics(testDogStatsdConfig())
+	if err != nil {
+		b.Error(err)
```

Comment:
b.Error marks the function as failed but does not immediately exit the goroutine (i.e. runs Fail, not FailNow)

Reply:
maybe just use `require.NoError(t, err)`?

Reply:
can you have both T and B in a single test? Couldn't find anything allowing that. Using 
```
		b.Log(err)
		b.FailNow()
```
for now

still curious about the statsd error

Reply:
sorry, `require.NoError(b, err)`. When you create functions that you want to use in both tests and benchmarks, then you usually have use the `testing.TB` interface as a parameter.

The `stretchr/testify` package uses an even more paired down interface (ie. `require.TestingT`) that both `*testing.T` and `*testing.B` conform to. So you can do `require.NoError` within both tests and benchmarks.

Reply:
very nice :)